### DISCUSSION TITLE
How to use fl_ctx across classes

### DISCUSSION THREAD

**Discussion by jingyunc** [NVIDIA/NVFlare#2099] [Q&A] [answered]
I have trouble using fl_ctx in the Swarm Learning controller (nvflare/app_common/ccwf/swarm_client_ctl.py). I wanted to use fl_ctx to pass aggr info from SwarmClientController._scatter (swarm_client_ctl.py:376) to  Gatherer.aggregate (swarm_client_ctl.py:161). 

At swarm_client_ctl.py:376, if I assign aggr to **new** prop of fl_ctx, like:
`fl_ctx.set_prop(CURRENT_AGGR, aggr, private=False, sticky=True)`
Then I got an error "NameError: name 'CURRENT_AGGR' is not defined"

But If I use **existing** prop name, like:
`fl_ctx.set_prop(AppConstants.DATA_CLIENT, aggr, private=False, sticky=True)`
And then retrieve the prop at swarm_client_ctl.py:161, like:
`aggr = fl_ctx.get_prop(AppConstants.DATA_CLIENT)`
The retrieved aggr was "None"

Can anyone advise on how to do this right? Thanks.

**Comment by SYangster** [comment, marked as answer]
When setting a new prop of fl_ctx, there should be no issues with this, is `CURRENT_AGGR` a str?

Retrieving the public prop from the Gatherer requires an additional step given that setting the prop and getting the prop can occur across different sites (since the scattering client can be different from the gathering client).

The FL Context contains a peer context that contains public props when processing a message from another party (https://nvflare.readthedocs.io/en/main/programming_guide/fl_context.html#peer-context-fl-ctx-get-peer-context). However a peer context is only available in certain situations for cross-component data sharing (eg. certain callback functions, execute(), and handle_event() etc). In swarm_client_ctl.py:161, the fl_ctx of the Gatherer does contain a peer context (if you trace the fl_ctx all the way back, it comes from client_ctl.py:375 which is in execute()). Thus we can use `get_peer_context()` at 161 and get the public prop we set in `_scatter()`.

Lastly, for this specific example, the `aggr` client would be the client doing `Gatherer.aggregrate()`, so this information isn’t needed.

**Comment by YuanTingHsieh** [comment]
Add to @SYangster 's point,
@jingyunc you need to either first declare the variable CURRENT_AGGR or just use a string

So:
```
CURRENT_AGGR = "CURRENT_AGGR"
fl_ctx.set_prop(CURRENT_AGGR, aggr, private=False, sticky=True)
```

Or:
```fl_ctx.set_prop("CURRENT_AGGR", aggr, private=False, sticky=True)```

**Comment by jingyunc** [comment]
Thanks @SYangster and @YuanTingHsieh . My issue was solved by both of your suggestions combined.
